### ISSUE TITLE
`no-plusplus` rule errors with valid JS code, requires more code as workaround

### ISSUE THREAD

**Issue by kureci** [eslint/eslint#11343]
**What rule do you want to change?**
`no-plusplus`

**Does this change cause the rule to produce more or fewer warnings?**
fewer, depends on approach taken to solve it

**How will the change be implemented? (New option, new default behavior, etc.)?**
I'm here to suggest a change in a high-level, haven't looked at the source code to suggest a solution implementation.

**Please provide some example code that this change will affect:**
```js
let a = 1;
let b;
let c = b || a++;
```

**What does the rule currently do for this code?**
At the moment the above throws a violation of the `no-plusplus` rule. The only workaround for satisfying the rule is to increment `a` in a new statement to have the same effect, because:
`let c = b || a+=1;` (which is the recommended solution from the docs) throws `Uncaught ReferenceError: Invalid left-hand side in assignment`, and
`let c = b || (a+1);` increments `a` **before** assigning to `c`, which isn't the intention here.

**What will the rule do after it's changed?**
It should either allow  valid statements like the above or only throw a warning instead of an error

**Are you willing to submit a pull request to implement this change?**
Possibly, haven't looked at the source code yet.

**Comment by g-plane** [comment]
Hi @kureci , thanks for opening this issue.

I think ESLint works as intended. Actually, you should add parens like this: `let c = b || (a += 1)`. 

Also, I will consider updating the docs.

**Comment by kureci** [comment]
@g-plane `c = b || (a += 1)` is **not** the same as `c = b || a++`
`(a+=1)` will increment `a` and then assign to `c` - i.e. pre-increment
`a++` will be a post-increment. See https://en.wikipedia.org/wiki/Increment_and_decrement_operators

So the two will have different results

**Comment by not-an-aardvark** [comment]
I think advocates for the `no-plusplus` rule would argue that code like `c = b || a++` is confusing to read. The code is "valid" in that it doesn't cause a runtime error, but it's the type of thing that (one might argue) you shouldn't use. It seems like every piece of code reported by `no-plusplus` is similarly "valid" but potentially confusing, so it might be better to simply disable the `no-plusplus` rule if you want to use code like that.

**Comment by kureci** [comment]
I agree with your comment @not-an-aardvark - but [https://github.com/eslint/eslint/pull/11387](PR#11387) which was raised to supposedly clarify this is just wrong. Feel free to check the PR and read my comment on there too.

I was just hoping that I wouldn't need to disable the rule to cater for post-increment statements, which are very well-known, long-used statements in most programming languages. Unlike placings of parens or braces which a lot of people have opinions on.

**Comment by platinumazure** [comment]
@kureci I'm curious why you're even using the rule. If you want to be able to use statements like `let c = b || a++;`, then what statements are you trying to block with the rule?

It really seems like you just don't need this rule, which is totally okay. ESLint is designed to be configurable and no rule is "required" in any setup.

**Comment by kureci** [comment]
@platinumazure you're right indeed, I don't need to be using this rule. It's just because I'm using Airbnb's ESLint as the base, and therefore now need to 'disable' the rule as it's enabled in the base.

Closing this issue.
